Mage.Sets/src/mage/cards/k/KayaSpiritsJustice.java

```diff
@@ -117,12 +104,52 @@ public boolean checkEventType(GameEvent event, Game game) {
         return event.getType() == GameEvent.EventType.ZONE_CHANGE_BATCH;
     }
 
+    @Override
+    public Stream<ZoneChangeEvent> filterBatchEvent(GameEvent event, Game game) {
+        // From Battlefield
+        Stream<ZoneChangeEvent> filteredBattlefield = ((ZoneChangeBatchEvent) event)
+                .getEvents()
+                .stream()
+                .filter(e -> e.getFromZone() == Zone.BATTLEFIELD)
+                .filter(e -> e.getToZone() == Zone.EXILED)
+                .filter(e -> Optional
+                        .of(e)
+                        .map(ZoneChangeEvent::getTargetId)
+                        .map(game::getCard)
+                        .filter(card -> {
+                            Permanent permanent = game.getPermanentOrLKIBattlefield(card.getId());
+                            return StaticFilters.FILTER_PERMANENT_CREATURE
+                                    .match(permanent, getControllerId(), this, game);
+                        })
+                        .isPresent()
```

Comment:
this looks wrong - the isPresent() will return true whether the Boolean is true or false, as long as the permanent is not null, which I don't think is what you want.

Reply:
Oh right, need `FILTER_CONTROLLED_PERMANENT_CREATURE`.

This is not about the Optional::filter logic right?

Reply:
No, I didn't notice the controller aspect. I haven't tested but I'm pretty sure your Optional usage won't work as intended.

Reply:
Isn't that the expected behavior in Java 8?
```
System.out.println(Optional.of(2).filter(n > 1).isPresent()); // true
System.out.println(Optional.of(0).filter(n > 1).isPresent()); // false
```

Reply:
![image](https://github.com/magefree/mage/assets/34709007/76f59504-6fb0-4e0d-bb7e-8a49d2f01b17)

Reply:
My apologies then, guess I have to go back and read the docs

Reply:
There were multiple times with Optional misunderstand already (in all use cases it was used as alternative to “if not null then execute some code”). I don’t like it too.

Reply:
There's nothing wrong here, Optional is fine style (and often cleaner). I was reading too quickly and made a faulty assumption.